kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SignalManager.java

```diff
@@ -0,0 +1,194 @@
+              + signalDirectory, e);
+    }
+  }
+
+  /**
+   * Loads a signal manager that stores metadata data under the given signal
+   * metadata directory if it exists. Returns <code>null</code> if it does not.
+   *
+   * @param conf the Hadoop configuration
+   * @param signalDirectory directory in which the manager stores signals.
+   *
+   * @return a signal manager instance, or <code>null</code> if the given
+   * directory does not exist.
+   */
+  public static SignalManager load(Configuration conf, Path signalDirectory) {
+    try {
+      FileSystem rootFileSystem = signalDirectory.getFileSystem(conf);
+
+      if (rootFileSystem.exists(signalDirectory)) {
+        return new SignalManager(signalDirectory, rootFileSystem);
+      } else {
+        return null;
+      }
+    } catch (IOException e) {
+      throw new DatasetIOException ("Cannot load signal manager at:"
+              + signalDirectory, e);
+    }
+  }
+
+  private SignalManager(Path signalDirectory, FileSystem rootFileSystem)
+          throws IOException  {
+    this.signalDirectory = signalDirectory;
+    this.rootFileSystem = rootFileSystem;
+  }
+
+  /**
+   * Create a signal for the specified constraints.
+   *
+   * @param viewConstraints The constraints to create a signal for.
+   *
+   * @throws DatasetException if the signal could not be created.
+   */
+  public void signalReady(Constraints viewConstraints) {
+    String normalizedConstraints = getNormalizedConstraints(viewConstraints);
+
+    Path signalPath = new Path(signalDirectory, normalizedConstraints);
+    try{
+      // create the output stream to overwrite the current contents, if the directory or file
+      // exists it will be overwritten to get a new timestamp
+      FSDataOutputStream os = rootFileSystem.create(signalPath, true);
+      os.close();
+    } catch (IOException e) {
+      throw new DatasetIOException("Could not access signal path: " + signalPath, e);
+    }
+  }
+
+  /**
+   * Check the last time the specified constraints have been signaled as ready.
+   *
+   * @param viewConstraints The constraints to check for a signal.
+   *
+   * @return the timestamp of the last time the constraints were signaled as ready.
+   *          if the constraints have never been signaled, -1 will be returned.
+   *
+   * @throws DatasetException if the signals could not be accessed.
+   */
+  public long getReadyTimestamp(Constraints viewConstraints) {
+    String normalizedConstraints = getNormalizedConstraints(viewConstraints);
+
+    Path signalPath = new Path(signalDirectory, normalizedConstraints);
+    // check if the signal exists
+    try {
+      try {
+        FileStatus signalStatus = rootFileSystem.getFileStatus(signalPath);
+        return signalStatus.getModificationTime();
+      } catch (final FileNotFoundException ex) {
+        // empty, will be thrown when the signal path doesn't exist
+      }
+      return -1;
+    } catch (NumberFormatException e) {
```

Comment:
Good catch 8ae9928, this was around in the previous implementation of the signal manager that made use of the FileContext API so we could get finer grained signal times than the modified timestamps themselves.